agent/consul/autopilot/autopilot.go

```diff
@@ -0,0 +1,457 @@
+package autopilot
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/consul/agent/metadata"
+	"github.com/hashicorp/raft"
+	"github.com/hashicorp/serf/serf"
+)
+
+// Delegate is the interface for the Autopilot mechanism
+type Delegate interface {
+	FetchStats(ctx context.Context, servers []*metadata.Server) map[string]*ServerStats
+	GetOrCreateAutopilotConfig() (*Config, bool)
+	NumPeers() (int, error)
+	PromoteNonVoters(*Config, OperatorHealthReply) ([]raft.Server, error)
+	Raft() *raft.Raft
+	Serf() *serf.Serf
+}
+
+// Autopilot is a mechanism for automatically managing the Raft
+// quorum using server health information along with updates from Serf gossip.
+// For more information, see https://www.consul.io/docs/guides/autopilot.html
+type Autopilot struct {
+	logger          *log.Logger
+	delegate        Delegate
+	validServerFunc func(serf.Member) bool
+
+	interval       time.Duration
+	healthInterval time.Duration
+
+	clusterHealth     OperatorHealthReply
+	clusterHealthLock sync.RWMutex
+
+	removeDeadCh chan struct{}
+	shutdownCh   chan struct{}
+	waitGroup    sync.WaitGroup
+}
+
+func NewAutopilot(logger *log.Logger, delegate Delegate, serverFunc func(serf.Member) bool, interval, healthInterval time.Duration) *Autopilot {
+	return &Autopilot{
+		logger:          logger,
+		delegate:        delegate,
+		validServerFunc: serverFunc,
+		interval:        interval,
+		healthInterval:  healthInterval,
+	}
+}
+
+func (a *Autopilot) Start() {
+	a.removeDeadCh = make(chan struct{})
```

Comment:
I put these here so they'd be reset in case autopilot was started/stopped multiple times on the same node from gaining/losing leadership repeatedly. `removeDeadCh` was ok to be moved to `NewAutopilot()` though, so I did that